pyproject.toml

```diff
@@ -1,7 +1,80 @@
 [build-system]
-requires = [ "setuptools", "wheel" ]
+requires = ["setuptools >= 63.0.0"]
 build-backend = "setuptools.build_meta"
 
+[project]
+name = "bitsandbytes"
+dynamic = ["version"]
+description = "k-bit optimizers and matrix multiplication routines."
+authors = [{name="Tim Dettmers", email="dettmers@cs.washington.edu"}]
+requires-python = ">=3.8"
+readme = "README.md"
+license = {file="LICENSE"}
+keywords = [
+    "gpu",
+    "optimizers",
+    "optimization",
+    "8-bit",
+    "quantization",
+    "compression"
+]
+classifiers = [
+    "Development Status :: 4 - Beta",
+    "License :: OSI Approved :: MIT License",
+    "Environment :: GPU :: NVIDIA CUDA :: 11",
+    "Environment :: GPU :: NVIDIA CUDA :: 12",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Science/Research",
+    "Operating System :: POSIX :: Linux",
+    "Operating System :: MacOS",
+    "Operating System :: Microsoft :: Windows",
+    "Programming Language :: C++",
+    "Programming Language :: Python :: Implementation :: CPython",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Topic :: Scientific/Engineering :: Artificial Intelligence"
+]
+dependencies = [
+    "torch",
```

Comment:
Don't we need any version constraint at all? Maybe at least a minimum version?

Reply:
I do think we should have at least minimums, but I'm not sure what they should be. I took these runtime dependencies out of `setup.py` just as they were.

If it were up to me, I would pin `torch>=1.13.0` on the low end for all platforms/backends assuming everything works. Why?

* Oldest release which did not ship any wheels with CUDA < 11
  * Migrated to CUDA 11.6 and 11.7, which is right in line where I think bitsandbytes cutoff should be [1]
  * Apple silicon support (MPS) promoted from experimental/prototype to beta
    * 🤗 [Accelerate docs](https://huggingface.co/docs/accelerate/en/usage_guides/mps) "strongly recommend" torch>=1.13 for MPS
* AWS has [HuggingFace containers](https://github.com/aws/deep-learning-containers/blob/master/available_images.md#huggingface-training-containers)with the oldest having `torch==1.13.1+cu117` on py3.9
** In general, all of the Python 3.9+ containers with PyTorch appear to be torch>=1.13.0
* [Azure Container for PyTorch](https://learn.microsoft.com/en-us/azure/machine-learning/resource-azure-container-for-pytorch?view=azureml-api-2) still has a 1.13. configuration

But it could also be a little more permissive for some environments; e.g. transformers recently adopted a constraint of `torch>=1.11,!=1.12.0` in huggingface/transformers#28207, which was also adopted in UKPLab/sentence-transformers#2432.

Reply:
IMHO (and I recognize this might ba a bit controversial) we should set minimum to what we test on, then pin to these when we test. I don't think we should overstretch to keep supporting old versions of dependencies at the expense of stability or complexity. So I would say ideally we set minimum to what we currently build on and either no upper bound or < next major.

As a user of any package I prefer to know what works or not and not having to be the guinea pig.

But this is my personal opinion, not sure what the rest of you think.

Reply:
Ideally we would have a CI setup that could test a wider matrix so we can be confident about how we pin these, for sure.

For me part of the point of this `bitsandbytes` project is to try and make things accessible for a wide range of users - so that's across hardware, software, and user's expertise. Since many people, particularly in academia, don't have much control over things like OS, python version, or even torch version, we've got to support a large range in order to make things easy for users. That does come at the cost of maintenance, obviously. Need to find a good balance here.

So at least with this change it's being restricted to `torch>=1.11,!=1.12.0`, which is in line with `transformers`. I don't know that all of those versions of torch will work, but it's an incremental step forward compared to `install_requires=["torch", "numpy"]`, and given the lack of issues around torch versions it does seem likely that it would work.